File: apps/autocad/code-app/src/__tests__/taxonomy/taxonomyMatrix.ts
/**
 * Taxonomy matrix — the single source of truth for reservation type ×
 * document subtype behaviour (docs/drawing-document-subtype-CONTRACT.md).
 *
 * This fixture is intentionally duplicated (same rows, same shape) in:
 *   - apps/code-app/src/__tests__/taxonomy/taxonomyMatrix.ts (this file)
 *   - solution/plugins/IssueNumbers.Tests/TaxonomyMatrix.cs
 *   - solution/scripts/tests/taxonomy_matrix.py
 * Keep all three in sync when the taxonomy changes.
 */

export type LibraryPair = "Drawing" | "Document";

export interface TaxonomyMatrixRow {
  /** enmax_acdnreservationtype option value. */
  reservationType: number;
  /** enmax_acdndocumentsubtype option value. */
  documentSubtype: number;
  /** Display name, matches the option set label. */
  label: string;
  /** Whether the base record produces numbered child sheets (-SSS) shown in lists. */
  createsChildren: boolean;
  /** Whether a base PDF is generated/expected for the base record itself. */
  basePdf: boolean;
  /** Which SharePoint library pair (Drawing* vs Document*) this subtype uses. */
  libraryPair: LibraryPair;
  /** Whether "Existing" sequence type (append to an existing base) is allowed. */
  existingAllowed: boolean;
  /** Default value of the per-taxonomy checkout enable toggle. */
  checkoutDefault: boolean;
  /** Whether numbered children may be appended to an existing base after creation. */
  appendAllowed: boolean;
}

export const TAXONOMY_MATRIX: readonly TaxonomyMatrixRow[] = [
  {
    reservationType: 1,
    documentSubtype: 1,
    label: "Drawing Document",
    createsChildren: false,
    basePdf: true,
    libraryPair: "Drawing",
    existingAllowed: false,
    checkoutDefault: true,
    appendAllowed: false,
  },
  {
    reservationType: 1,
    documentSubtype: 2,
    label: "Drawing",
    createsChildren: true,
    basePdf: true,
    libraryPair: "Drawing",
    existingAllowed: true,
    checkoutDefault: true,
    appendAllowed: true,
  },
  {
    reservationType: 2,
    documentSubtype: 3,
    label: "Standard",
    createsChildren: false,
    basePdf: true,
    libraryPair: "Document",
    existingAllowed: false,
    checkoutDefault: true,
    appendAllowed: false,
  },
  {
    reservationType: 2,
    documentSubtype: 4,
    label: "Procedure",
    createsChildren: true,
    basePdf: true,
    libraryPair: "Document",
    existingAllowed: false,
    checkoutDefault: true,
    appendAllowed: false,
  },
  {
    reservationType: 2,
    documentSubtype: 5,
    label: "Form",
    createsChildren: true,
    basePdf: false,
    libraryPair: "Document",
    existingAllowed: true,
    checkoutDefault: true,
    appendAllowed: true,
  },
] as const;

export function taxonomyMatrixRow(
  reservationType: number,
  documentSubtype: number,
): TaxonomyMatrixRow | undefined {
  return TAXONOMY_MATRIX.find(
    (row) => row.reservationType === reservationType && row.documentSubtype === documentSubtype,
  );
}

File: apps/autocad/solution/scripts/tests/taxonomy_matrix.py
"""Taxonomy matrix — the single source of truth for reservation type x
document subtype behaviour (docs/drawing-document-subtype-CONTRACT.md).

This fixture is intentionally duplicated (same rows, same shape) in:
  - apps/code-app/src/__tests__/taxonomy/taxonomyMatrix.ts
  - solution/plugins/IssueNumbers.Tests/TaxonomyMatrix.cs
  - solution/scripts/tests/taxonomy_matrix.py (this file)
Keep all three in sync when the taxonomy changes.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class TaxonomyMatrixRow:
    reservation_type: int
    document_subtype: int
    label: str
    creates_children: bool
    base_pdf: bool
    library_pair: str  # "Drawing" | "Document"
    existing_allowed: bool
    checkout_default: bool
    append_allowed: bool


TAXONOMY_MATRIX: tuple[TaxonomyMatrixRow, ...] = (
    TaxonomyMatrixRow(
        reservation_type=1,
        document_subtype=1,
        label="Drawing Document",
        creates_children=False,
        base_pdf=True,
        library_pair="Drawing",
        existing_allowed=False,
        checkout_default=True,
        append_allowed=False,
    ),
    TaxonomyMatrixRow(
        reservation_type=1,
        document_subtype=2,
        label="Drawing",
        creates_children=True,
        base_pdf=True,
        library_pair="Drawing",
        existing_allowed=True,
        checkout_default=True,
        append_allowed=True,
    ),
    TaxonomyMatrixRow(
        reservation_type=2,
        document_subtype=3,
        label="Standard",
        creates_children=False,
        base_pdf=True,
        library_pair="Document",
        existing_allowed=False,
        checkout_default=True,
        append_allowed=False,
    ),
    TaxonomyMatrixRow(
        reservation_type=2,
        document_subtype=4,
        label="Procedure",
        creates_children=True,
        base_pdf=True,
        library_pair="Document",
        existing_allowed=False,
        checkout_default=True,
        append_allowed=False,
    ),
    TaxonomyMatrixRow(
        reservation_type=2,
        document_subtype=5,
        label="Form",
        creates_children=True,
        base_pdf=False,
        library_pair="Document",
        existing_allowed=True,
        checkout_default=True,
        append_allowed=True,
    ),
)


def find_row(reservation_type: int, document_subtype: int) -> TaxonomyMatrixRow | None:
    for row in TAXONOMY_MATRIX:
        if row.reservation_type == reservation_type and row.document_subtype == document_subtype:
            return row
    return None
